---
title: "Thunderbio scRNA-seq Report"
date: "`r Sys.Date()`"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    orientation: rows
    theme: 
      bootswatch: cosmo
      version: 5
params:
  sampleName : "sampleName"
  qualimap_out : "rnaseq_qc_results.txt"
  qualimap_gene_coverage : "qualimap_gene_coverage.txt"
  starsolo_matrixDir: "sample/Gene/filtered"
  starsolo_out : "Summary.csv"
  starsolo_bc : "UMIperCellSorted.txt"
  saturation_json: "saturation_out.json"
  nCPUs: 10
  nMem: 10000000000.0
  nFeature_RNA_lower: 200
  percent_mt_cutoff: 100
  pca_dim: 30
  resolution: 0.8
---

```{r include=FALSE}
library(tidyverse)
library(scales)
library(flexdashboard)
library(jsonlite)
library(DT)
##library(htmlwidgets)
library(plotly)
library(kableExtra)
library(Seurat)
library(future)
library(SeuratDisk)

set.seed(1234)
plan(multisession, workers = as.numeric(params$nCPUs))
options(future.globals.maxSize = as.numeric(params$nMem), sass.cache = FALSE)

## Rmd options
knitr::opts_chunk$set(
  comment = '', fig.retina = 4,
  warning = FALSE, message = FALSE
)


config_plotly_fig <- function(fig){
    config(
        fig,
        displaylogo = FALSE, 
        modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
                                   'hoverClosestCartesian', 'hoverCompareCartesian'),
        toImageButtonOptions = list(height= NULL, width= NULL, scale= 2)
    )
}

starsolo_summary <- read_csv(params$starsolo_out,
                             col_names = c("item", "value")) %>%
                             mutate(
                                 value = case_when(
                                 value > 1 ~ comma(value, accuracy = 1),
                                 value <=1 ~ percent(value)
                                 )
                             ) %>%
    mutate(item=str_replace_all(item, "GeneFull", "Gene"))
exon_cmd <- paste0("grep 'exonic' ", params$qualimap_out, " | awk '{print $NF}' | sed 's/(//; s/)//'")
exon_ratio <- system(exon_cmd, intern = TRUE)

intron_cmd <- paste0("grep 'intron' ", params$qualimap_out, " | awk '{print $NF}' | sed 's/(//; s/)//'")
intron_ratio <- system(intron_cmd, intern = TRUE)
```

```{r process_starsolo, include = FALSE}
STARSolo <- Read10X(data.dir=file.path(params$starsolo_matrixDir))
seuratData <- CreateSeuratObject(counts=STARSolo,
                             min.cells=0,
                             min.features=0, project=params$sampleName)
seuratData[['run']]=params$sampleName
seuratData[["percent.mt"]]=PercentageFeatureSet(object=seuratData, pattern = "MT-|mt-")

## save a raw loom file
seuratData.loom <- SaveLoom(
    seuratData, 
    filename = paste0(params$sampleName, ".raw.loom"), 
    overwrite = TRUE,
    verbose = FALSE
)

## Plot the genes counts, RNAs counts and mitochondrial proportions.
qcPlot <- VlnPlot(object=seuratData, features=c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3)

## Filter by mitochondrial percentage
seuratData <- subset(
    seuratData, 
    nFeature_RNA > as.numeric(params$nFeature_RNA_lower) & percent.mt < as.numeric(params$percent_mt_cutoff)
)

filtered_cells <- seuratData@meta.data %>% nrow
## Only perform clustering when more than 150 cells detected
if(filtered_cells >= 150){
    ## Normalization and scaling
    seuratData <- NormalizeData(object=seuratData)
    seuratData <- FindVariableFeatures(seuratData)
    seuratData <- ScaleData(seuratData, vars.to.regress=c("nCount_RNA", "percent.mt"), features=rownames(seuratData))
    
    seuratData <- RunPCA(object=seuratData, npcs=100)
    
    ## set pca_dim to be used, 30
    pca_dim <- as.numeric(params$pca_dim)
    ## set resolution
    resolution <- as.numeric(params$resolution)
    
    seuratData <- FindNeighbors(object=seuratData, reduction="pca", dims=1:pca_dim)
    seuratData <- FindClusters(object=seuratData, resolution=resolution)
    seuratData <- RunUMAP(object=seuratData, reduction="pca", dims=1:pca_dim)
    
    ##Idents(object=seuratData) <- paste("RNA_snn_res.", resolution, sep="")
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
}
```

```{css echo = FALSE}
.chart-title {
  font-size: 1.2rem;
  font-weight: 700;
}
```

Mapping Result {data-orientation=rows}
====================

<h2 style="margin-top: 0.5rem; font-weight: bold;">Sample: `r params$sampleName`</h2>

Row
----------

### Estimated Number of Cells

```{r}
cellNum <- starsolo_summary %>% 
    filter(item == "Estimated Number of Cells") %>% 
    pull(value)
valueBox(cellNum, icon = "fa-circle-notch", color = "info")
```

### Mean Reads per Cell

```{r}
readsMean <- starsolo_summary %>% 
    filter(item == "Mean Reads per Cell") %>% 
    pull(value)
valueBox(readsMean, icon = "fa-chart-bar", color = "primary")
```

### Median Genes per Cell

```{r}
medianGenePerCell <- starsolo_summary %>% 
    filter(str_detect(item, regex("Median Gene(Full)* per Cell"))) %>% 
    pull(value)
valueBox(medianGenePerCell, icon = "fa-barcode", color = "warning")
```

Row
----------

### Cell Stats

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Fraction of Unique Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
names(d) <- c("","")                       
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### Knee Plot

```{r}
umi_dist <- read_tsv(params$starsolo_bc, col_names =c("UMI"))
cellNum <- str_replace(cellNum, ",", "") %>% as.numeric()
umi_dist <-  umi_dist %>%
    mutate(
        cell = c(1:nrow(umi_dist)),
        group = case_when(
            cell <= cellNum ~ "cell",
            TRUE ~ "background"
            )
        )

fig <- plot_ly(data = umi_dist, x = ~cell, y = ~UMI, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'UMI: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3))
fig <- layout(fig, xaxis = list(type = "log"),yaxis = list(type = "log"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()

##ggplot(umi_dist, aes(x=cell, y=UMI, color = group)) + 
##    geom_point(alpha = 0.8) +
##    geom_vline(xintercept = cellNum, 
##               linetype = "dashed", color = "grey") +
##    scale_x_log10(n.breaks = 8, labels = scales::label_number_si()) + 
##    scale_y_log10(n.breaks = 5, labels = scales::label_number_si()) +
##    scale_color_manual(values =c("grey", "royalblue")) +
##    xlab("Cell") + ylab("UMI") +
##    theme_bw() + 
##    theme(panel.grid = element_blank(), 
##          axis.text = element_text(color="black"),
##          legend.position = "none")
```

Row
----------

### Sequencing Stats

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
                                        
```

### Mapping Stats

```{r}
d <- starsolo_summary %>% 
    filter(item %in% c("Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multipe Gene",
                       "Reads Mapped to GeneFull: Unique+Multipe GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = exon_ratio) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = intron_ratio)
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

Cell Cluster {data-orientation=rows}
====================

Row
----------

### Cell Cluster (colored by cell group)

```{r}
## Only plot cluster when there are more than 150 filtered cells
if(filtered_cells >= 150){
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
    ## extract data from seuratData
    umap_data <- Embeddings(object = seuratData[["umap"]])[colnames(seuratData), c(1, 2)]
    clusterData <- merge(umap_data, seuratData@meta.data, by=0) %>% 
        dplyr::rename("bc" = `Row.names`)
    cellCount <- clusterData %>% group_by(seurat_clusters) %>% summarise(cellCount = n())
    totalCell <- cellCount %>% pull(cellCount) %>% sum()
    clusterData <- clusterData %>% left_join(cellCount, by = "seurat_clusters") %>%
        mutate(totalCell=totalCell,
               text = paste0(cellCount, "/", totalCell))
    
    plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~seurat_clusters, 
            text = ~text,
            type = 'scatter', mode = 'markers',
            colors = 'Dark2',
            hovertemplate = 'cells: %{text}<extra><b>Cluster: %{fullData.name}</b></extra>') %>%
        config_plotly_fig() %>%
        toWebGL()
    
    ##clusterPlot %>% ggplotly() %>%
    ##    config(displaylogo = FALSE,
    ##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
    ##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

### Cell Cluster (colored by read depth)


```{r}
if(filtered_cells >= 150){
clusterData <- clusterData %>%
    mutate(log10_UMI = log10(nCount_RNA))
    
plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~log10_UMI,  text = ~nCount_RNA,
        type = 'scatter', mode = 'markers',
        hovertemplate = 'UMI: %{text}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()

##depthPlot <- FeaturePlot(object = seuratData, features = 'log_nCount_RNA') + scale_color_continuous(name="log10 Read Counts", type = "viridis", direction = 1) + labs(title = NULL)
##depthPlot %>% ggplotly() %>%
##    config(displaylogo = FALSE, 
##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

Row {style="margin-right: 8px"}
----------

### Marker Genes of Clusters (top5) {style="width: 100% !important"}

```{r}
if(filtered_cells >= 150){
    markerGenes <- FindAllMarkers(seuratData, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
    markerGenes %>%
        group_by(cluster) %>%
        slice_max(n = 5, order_by = avg_log2FC) %>% 
        dplyr::relocate(gene, .before=1) %>%
        mutate(p_val = signif(p_val, 3),
               avg_log2FC = signif(avg_log2FC, 3),
               p_val_adj = signif(p_val_adj, 3)) %>%
        DT::datatable(rownames = FALSE, width = "100%",
                      options = list(scrollX = T, scrollY = F, pageLength = 10))
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

Row
----------

### Cell Metrics

```{r}
qcPlot
```

### Gene Coverage

```{r}
read_tsv(params$qualimap_gene_coverage) %>%
    ggplot(aes(x=`#Transcript position`, y=`Transcript coverage profile`)) +
    geom_line(size=1, color = "#551a8b", alpha=0.8) +
    scale_y_continuous(labels = label_comma()) +
    theme_bw() + theme(panel.grid = element_blank())
```


Row
----------

### Sequencing Saturation

```{r}
saturation_data <- fromJSON(params$saturation_json) %>%
    mutate(reads=as.numeric(reads),
           saturation=as.numeric(saturation),
           meanReadPerCell=as.numeric(meanReadPerCell),
           medianGenePerCell=as.numeric(medianGenePerCell))
plot_ly(saturation_data, x = ~reads, y= ~saturation,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Read depth: %{x}<br>Saturation: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

### Median Genes per Cell

```{r}
plot_ly(saturation_data, x = ~meanReadPerCell, y= ~medianGenePerCell,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Mean read per cell: %{x}<br>Median gene per cell: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

```{r include=FALSE}
starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Fraction of Unique Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected",
                       "Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multipe Gene",
                       "Reads Mapped to GeneFull: Unique+Multipe GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = exon_ratio) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = intron_ratio) %>%
    mutate(sampleID=params$sampleName) %>% 
    pivot_wider(sampleID, names_from = item, values_from= value) %>% 
    write_json(paste0(params$sampleName, ".metrics.json"), pretty=TRUE)
```
