---
title: "ThunderBio VDJ Report"
date: "`r Sys.Date()`"
output: 
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    orientation: rows
    theme: 
      bootswatch: flatly
      version: 5
params:
  sampleName : "sampleName"
  qualimap_out : "rnaseq_qc_results.txt"
  qualimap_gene_coverage : "qualimap_gene_coverage.txt"
  starsolo_matrixDir: "sample/Gene/filtered"
  starsolo_out : "Summary.csv"
  starsolo_bc : "UMIperCellSorted.txt"
  saturation_json: "saturation_out.json"
  version_json: "versions.json"
  nCPUs: 10
  nMem: 10000000000.0
  nFeature_RNA_lower: 200
  percent_mt_cutoff: 100
  pca_dim: 30
  resolution: 0.8
  VDJ_B_report: "VDJ_B_report.tsv"
  VDJ_B_airr: "VDJ_B_airr.tsv"
  VDJ_B_kneeOut: "VDJ_B_kneeOut.tsv"
  VDJ_B_finalOut: "VDJ_B_finalOut.tsv"
  VDJ_B_cells: "VDJ_B_cells.tsv"
  VDJ_B_metrics: "VDJ-B_metrics.json"
  VDJ_B_cloneType : "VDJ-B.cloneType.tsv"
  VDJ_T_report: "VDJ_T_report.tsv"
  VDJ_T_airr: "VDJ_T_airr.tsv"
  VDJ_T_kneeOut: "VDJ_T_kneeOut.tsv"
  VDJ_T_finalOut: "VDJ_T_finalOut.tsv"
  VDJ_T_cells: "VDJ_T_cells.tsv"
  VDJ_T_metrics: "VDJ-T_metrics.json"
  VDJ_T_cloneType: "VDJ-T.cloneType.tsv"
---

```{r include=FALSE}
library(tidyverse)
library(scales)
library(flexdashboard)
library(jsonlite)
library(DT)
##library(htmlwidgets)
library(plotly)
library(kableExtra)
library(Seurat)
library(future)
library(SeuratDisk)
##library(VennDiagram)

set.seed(1234)
plan(multisession, workers = as.numeric(params$nCPUs))
options(future.globals.maxSize = as.numeric(params$nMem), sass.cache = FALSE)
RhpcBLASctl::blas_set_num_threads(1) # https://github.com/satijalab/seurat/issues/3991

## Rmd options
knitr::opts_chunk$set(
  comment = '', fig.retina = 4,
  warning = FALSE, message = FALSE
)


config_plotly_fig <- function(fig, ...){
    config(
        fig,
        displaylogo = FALSE, 
        modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
                                   'hoverClosestCartesian', 'hoverCompareCartesian'),
        toImageButtonOptions = list(height= NULL, width= NULL, scale= 2),
        ...
    )
}

## define info icon funciton
infoIcon <- function(info,
                     class = "fa fa-info-circle",
                     style="font-size: 1.25rem; color: #545b62;"){
  ## Make HTML
  outTag <- paste0('<i class="',
                   class,
                   '" aria-hidden="true" ',
                   'data-tippy-content="',
                   info,
                   '" ',
                   'style="',
                   style,'"></i>')
  return(HTML(outTag))
}


## decide whether to incldue each page
## conditional page solution is from: https://stackoverflow.com/questions/59202693/how-can-i-create-a-conditional-flexdashboard-layout
if(is.null(params$starsolo_matrixDir) || params$starsolo_matrixDir == "") {
    include_gex <- FALSE
}else{
    include_gex <- TRUE
}
if(is.null(params$VDJ_T_metrics) || params$VDJ_T_metrics == "") {
    include_tcr <- FALSE
}else{
    include_tcr <- TRUE
}
if(is.null(params$VDJ_B_metrics) || params$VDJ_B_metrics == "") {
    include_bcr <- FALSE
}else{
    include_bcr <- TRUE
}
```

`r if(!include_gex){"\\begin{comment}"} else {NULL}`

```{r process_gex, include = FALSE, eval = include_gex}
is_number_string <- function(x){
    str_split(x, pattern="") %>% 
        unlist %>% str_detect("[0-9.]") %>% 
        all()
}
starsolo_summary <- read_csv(params$starsolo_out,
                             col_names = c("item", "value"),
                             col_types = cols("c", "c"))
starsolo_summary$data <- NA
for(i in 1:nrow(starsolo_summary)){
    if(is_number_string(starsolo_summary$value[i])){
        if(as.numeric(starsolo_summary$value[i]) >= 1){
            starsolo_summary$data[i] <- comma(as.numeric(starsolo_summary$value[i]), accuracy = 1)
        }else{
            starsolo_summary$data[i] <- percent(as.numeric(starsolo_summary$value[i]))
        }
    }else{
        starsolo_summary$data[i] <- starsolo_summary$value[i]
    }
}
starsolo_summary <- starsolo_summary %>%
    select(-value) %>%
    dplyr::rename("value" = "data") %>%
    mutate(item=str_replace_all(item, "GeneFull", "Gene"))
exon_cmd <- paste0("grep 'exonic' ", params$qualimap_out, " | awk '{print $NF}' | sed 's/(//; s/)//'")
exon_ratio <- system(exon_cmd, intern = TRUE)

intron_cmd <- paste0("grep 'intron' ", params$qualimap_out, " | awk '{print $NF}' | sed 's/(//; s/)//'")
intron_ratio <- system(intron_cmd, intern = TRUE)

STARSolo <- Read10X(data.dir=file.path(params$starsolo_matrixDir))
seuratData <- CreateSeuratObject(counts=STARSolo,
                             min.cells=0,
                             min.features=0, project=params$sampleName)
seuratData[['run']]=params$sampleName
seuratData[["percent.mt"]]=PercentageFeatureSet(object=seuratData, pattern = "MT-|mt-")

## save a raw loom file
SaveH5Seurat(
    seuratData, 
    filename = paste0(params$sampleName, ".raw.h5seurat"), 
    overwrite = TRUE,
    verbose = FALSE
)

## Plot the genes counts, RNAs counts and mitochondrial proportions.
qcPlot <- VlnPlot(object=seuratData, features=c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol=3, group.by="run")

## Filter by gene counts and mitochondrial percentage
## Check the number of cells left first, or subset will throw an error
filtered_cells <- seuratData@meta.data %>% 
    filter(nFeature_RNA > max(200, as.numeric(params$nFeature_RNA_lower)),
           percent.mt < as.numeric(params$percent_mt_cutoff)) %>% 
           nrow
message("filtered_cells is ", filtered_cells)
if(filtered_cells>0){
    seuratData <- subset(
        seuratData, 
        nFeature_RNA > as.numeric(params$nFeature_RNA_lower) & percent.mt < as.numeric(params$percent_mt_cutoff)
    )
    filtered_cells <- seuratData@meta.data %>% nrow
}

## Only perform clustering when more than 150 cells detected
if(filtered_cells >= 150){
    ## Normalization and scaling
    ##seuratData <- NormalizeData(object=seuratData)
    ##seuratData <- FindVariableFeatures(seuratData)
    ##seuratData <- ScaleData(seuratData, vars.to.regress=c("nCount_RNA", "percent.mt"), features=rownames(seuratData))
    
    ## replace above commands with sctransform
    ## default use top 3000 variable features
    seuratData <- SCTransform(seuratData, vars.to.regress = "percent.mt", conserve.memory = TRUE)
    
    seuratData <- RunPCA(object=seuratData, npcs=100)
    
    ## set pca_dim to be used, 30
    pca_dim <- as.numeric(params$pca_dim)
    ## set resolution
    resolution <- as.numeric(params$resolution)
    
    seuratData <- FindNeighbors(object=seuratData, reduction="pca", dims=1:pca_dim)
    seuratData <- FindClusters(object=seuratData, resolution=resolution)
    seuratData <- RunUMAP(object=seuratData, reduction="pca", dims=1:pca_dim)
    
    ##Idents(object=seuratData) <- paste("RNA_snn_res.", resolution, sep="")
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
}
```

```{css echo = FALSE}
.chart-title {
  font-size: 1.2rem;
  font-weight: 700;
  margin-left: 10px;
  margin-right: 10px;
}
```

```{js echo = FALSE}
// shrink navbar height
document.querySelector(".navbar").classList.add("pt-1")
document.querySelector(".navbar").classList.add("pb-1")
```

GEX {data-orientation=rows}
====================

<h2 style="margin-top: 0.5rem; font-weight: bold;">Sample: `r params$sampleName` (GEX)</h2>

Row
----------

### Estimated Number of Cells

```{r eval = include_gex}
cellNum <- starsolo_summary %>% 
    filter(item == "Estimated Number of Cells") %>% 
    pull(value)
valueBox(cellNum, icon = "fa-circle-notch", color = "info")
```

### Mean Reads per Cell

```{r eval = include_gex}
readsMean <- starsolo_summary %>% 
    filter(item == "Mean Reads per Cell") %>% 
    pull(value)
valueBox(readsMean, icon = "fa-chart-bar", color = "primary")
```

### Median Genes per Cell

```{r eval = include_gex}
medianGenePerCell <- starsolo_summary %>% 
    filter(str_detect(item, regex("Median Gene(Full)* per Cell"))) %>% 
    pull(value)
valueBox(medianGenePerCell, icon = "fa-barcode", color = "warning")
```

Row
----------

### Cell Stats

```{r eval = include_gex}
d <- starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Reads in Cells Mapped to Gene",
                       "Fraction of Unique Reads in Cells",
                       "Fraction of Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
names(d) <- c("","")                       
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### Knee Plot

```{r eval = include_gex}
umi_dist <- read_tsv(params$starsolo_bc, col_names =c("UMI"))
cellNum <- str_replace(cellNum, ",", "") %>% as.numeric()
umi_dist <-  umi_dist %>%
    mutate(
        cell = c(1:nrow(umi_dist)),
        group = case_when(
            cell <= cellNum ~ "cell",
            TRUE ~ "background"
            )
        )

fig <- plot_ly(data = umi_dist, x = ~cell, y = ~UMI, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'UMI: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log"),yaxis = list(type = "log"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()

##ggplot(umi_dist, aes(x=cell, y=UMI, color = group)) + 
##    geom_point(alpha = 0.8) +
##    geom_vline(xintercept = cellNum, 
##               linetype = "dashed", color = "grey") +
##    scale_x_log10(n.breaks = 8, labels = scales::label_number_si()) + 
##    scale_y_log10(n.breaks = 5, labels = scales::label_number_si()) +
##    scale_color_manual(values =c("grey", "royalblue")) +
##    xlab("Cell") + ylab("UMI") +
##    theme_bw() + 
##    theme(panel.grid = element_blank(), 
##          axis.text = element_text(color="black"),
##          legend.position = "none")
```

Row {data-height=400}
----------

### Sequencing Stats

```{r eval = include_gex}
d <- starsolo_summary %>% 
    filter(item %in% c("Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Total Gene Detected",
                       "Total GeneFull Detected"))
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
                                        
```

### Mapping Stats

```{r eval = include_gex}
d <- starsolo_summary %>% 
    filter(item %in% c("Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multiple Gene",
                       "Reads Mapped to GeneFull: Unique+Multiple GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = exon_ratio) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = intron_ratio)
names(d) <- c("","")
d %>% kbl() %>% kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

Row
----------

### Cell Cluster (colored by cell group)

```{r eval = include_gex}
## Only plot cluster when there are more than 150 filtered cells
if(filtered_cells >= 150){
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
    ## extract data from seuratData
    umap_data <- Embeddings(object = seuratData[["umap"]])[colnames(seuratData), c(1, 2)]
    clusterData <- merge(umap_data, seuratData@meta.data, by=0) %>% 
        dplyr::rename("bc" = `Row.names`)
    cellCount <- clusterData %>% group_by(seurat_clusters) %>% summarise(cellCount = n())
    totalCell <- cellCount %>% pull(cellCount) %>% sum()
    clusterData <- clusterData %>% left_join(cellCount, by = "seurat_clusters") %>%
        mutate(totalCell=totalCell,
               text = paste0(cellCount, "/", totalCell))
    
    plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~seurat_clusters, 
            text = ~text,
            type = 'scatter', mode = 'markers',
            colors = 'Dark2',
            hovertemplate = 'cells: %{text}<extra><b>Cluster: %{fullData.name}</b></extra>') %>%
        config_plotly_fig() %>%
        toWebGL()
    
    ##clusterPlot %>% ggplotly() %>%
    ##    config(displaylogo = FALSE,
    ##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
    ##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

### Cell Cluster (colored by read depth)

```{r eval = include_gex}
if(filtered_cells >= 150){
clusterData <- clusterData %>%
    mutate(log10_UMI = log10(nCount_RNA))
    
plot_ly(clusterData, x = ~UMAP_1, y= ~UMAP_2, color = ~log10_UMI,  text = ~nCount_RNA,
        type = 'scatter', mode = 'markers',
        hovertemplate = 'UMI: %{text}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()

##depthPlot <- FeaturePlot(object = seuratData, features = 'log_nCount_RNA') + scale_color_continuous(name="log10 Read Counts", type = "viridis", direction = 1) + labs(title = NULL)
##depthPlot %>% ggplotly() %>%
##    config(displaylogo = FALSE, 
##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

Row {style="margin-right: 8px"}
----------

### Marker Genes of Clusters (top5) {style="width: 100% !important"}

```{r eval = include_gex}
if(filtered_cells >= 150){
    markerGenes <- FindAllMarkers(seuratData, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
    if("cluster" %in% colnames(markerGenes)){
        markerGenes %>%
            group_by(cluster) %>%
            slice_max(n = 5, order_by = avg_log2FC) %>% 
            dplyr::relocate(gene, .before=1) %>%
            mutate(p_val = signif(p_val, 3),
                   avg_log2FC = signif(avg_log2FC, 3),
                   p_val_adj = signif(p_val_adj, 3)) %>%
            DT::datatable(rownames = FALSE, width = "100%",
                          options = list(scrollX = T, scrollY = F, pageLength = 10))
    }else{
        cat("No DEGs detected.\n")
    }
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

```{r include=FALSE, eval = include_gex}
if(filtered_cells >= 150 && ("cluster" %in% colnames(markerGenes))){
    markerGenes %>%
        group_by(cluster) %>%
        dplyr::relocate(gene, .before=1) %>%
        write_tsv(paste0(params$sampleName, "_DEG.tsv"))
}
```

Row
----------

### Cell Metrics

```{r eval = include_gex}
qcPlot
```

### Gene Coverage

```{r eval = include_gex}
read_tsv(params$qualimap_gene_coverage) %>%
    ggplot(aes(x=`#Transcript position`, y=`Transcript coverage profile`)) +
    geom_line(size=1, color = "#551a8b", alpha=0.8) +
    scale_y_continuous(labels = label_comma()) +
    theme_bw() + theme(panel.grid = element_blank())
```


Row
----------

### Sequencing Saturation

```{r eval = include_gex}
saturation_data <- fromJSON(params$saturation_json)
saturation_data <- saturation_data[["saturation_data"]] %>%
    as_tibble() %>%
    mutate(across(.col = everything(), .fns = as.numeric))

plot_ly(saturation_data, x = ~reads, y= ~saturation,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Read depth: %{x}<br>Saturation: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

### Median Genes per Cell

```{r eval = include_gex}
plot_ly(saturation_data, x = ~meanReadPerCell, y= ~medianGenePerCell,
        type = 'scatter', mode = 'lines',
        hovertemplate = 'Mean read per cell: %{x}<br>Median gene per cell: %{y}<extra></extra>') %>%
    config_plotly_fig() %>%
    toWebGL()
```

```{r include=FALSE, eval = include_gex}
starsolo_summary %>% 
    filter(item %in% c("Estimated Number of Cells",
                       "Unique Reads in Cells Mapped to Gene",
                       "Unique Reads in Cells Mapped to GeneFull",
                       "Fraction of Unique Reads in Cells",
                       "Mean Reads per Cell",
                       "Median Reads per Cell",
                       "UMIs in Cells",
                       "Mean UMI per Cell",
                       "Median UMI per Cell",
                       "Mean Gene per Cell",
                       "Mean GeneFull per Cell",
                       "Median Gene per Cell",
                       "Median GeneFull per Cell",
                       "Total Gene Detected",
                       "Total GeneFull Detected",
                       "Number of Reads",
                       "Reads With Valid Barcodes",
                       "Sequencing Saturation",
                       "Q30 Bases in CB+UMI",
                       "Q30 Bases in RNA read",
                       "Reads Mapped to Genome: Unique+Multiple",
                       "Reads Mapped to Genome: Unique",
                       "Reads Mapped to Gene: Unique+Multiple Gene",
                       "Reads Mapped to GeneFull: Unique+Multiple GeneFull",
                       "Reads Mapped to Gene: Unique Gene",
                       "Reads Mapped to GeneFull: Unique GeneFull")) %>%
    add_row(item = "Reads Mapped to Exonic Regions", value = exon_ratio) %>%
    add_row(item = "Reads Mapped to Intronic Regions", value = intron_ratio) %>%
    mutate(sampleID=params$sampleName) %>% 
    pivot_wider(sampleID, names_from = item, values_from= value) %>% 
    write_json(paste0(params$sampleName, ".metrics.json"), pretty=TRUE)
```

`r if(!include_gex) {"\\end{comment}"} else {NULL}`

`r if(!include_tcr) {"\\begin{comment}"} else {NULL}`

VDJ-T {data-orientation=rows}
====================

<h2 style="margin-top: 0.5rem; font-weight: bold;">Sample: `r params$sampleName` (VDJ-T)</h2>

```{r include=FALSE, eval = include_tcr}
vdj_t_cloneType <- read_tsv(params$VDJ_T_cloneType) %>%
    arrange(desc(CellCount))
vdj_t_cloneType <- vdj_t_cloneType %>%
    mutate(id = 1:nrow(vdj_t_cloneType),
           Frequency = scales::label_percent(0.01)(Frequency))
```

Row
----------

### Estimated Number of Cells


```{r include = FALSE, eval = include_tcr}
vdj_t_metrics <- fromJSON(params$VDJ_T_metrics)
## write metrics to tsv
fromJSON(params$VDJ_T_metrics) %>% 
    as.data.frame %>% t() %>% as.data.frame() %>% 
    rownames_to_column("term") %>% 
    write_tsv(paste0(params$sampleName, "_VDJ-T.metrics.tsv"), col_names=FALSE)
```

```{r eval = include_tcr}
valueBox(vdj_t_metrics$cells %>% as.numeric() %>% label_comma()(), icon = "fa-circle-notch", color = "info")
```

### Mean Reads per Cell

```{r eval = include_tcr}
valueBox(vdj_t_metrics$meanReadsPerCell %>% as.numeric() %>% round %>% label_comma()(), icon = "fa-chart-bar", color = "primary")
```

### Total Clonetypes Discovered

```{r eval = include_tcr}
valueBox(nrow(vdj_t_cloneType) %>% label_comma()(), icon = "fa-barcode", color = "warning")
```

Row
----------

### VDJ Stats

```{r eval = include_tcr}
##Estimated Number of Cells
##Reads in Cells Mapped to Gene
##Fraction of Reads in Cells
##Mean Reads per Cell
##Median Reads per Cell
##UMIs in Cells
##Mean UMI per Cell
##Median UMI per Cell
##Mean Gene per Cell
##Median Gene per Cell

fractionAssembled <- as.numeric(vdj_t_metrics$totalReadsInCell)/(as.numeric(vdj_t_metrics$totalRawReads)*as.numeric(vdj_t_metrics$validBCreads))
fractionAssembled <- as.character(fractionAssembled)

vdj_t_tbl <- tribble(
  ~term, ~value,
  "Estimated Number of Cells", vdj_t_metrics$cells,
  "Reads in Cells Mapped to VDJ Gene", vdj_t_metrics$totalReadsInCell,
  "Fraction of Reads Used in Assembly", fractionAssembled,
  "Mean Reads per Cell", vdj_t_metrics$meanReadsPerCell,
  "Median Reads per Cell", vdj_t_metrics$medianReadsPerCell,
  "Mean UMIs per Cell", vdj_t_metrics$meanUMIsPerCell,
  "Median UMIs per Cell", vdj_t_metrics$medianUMIsPerCell,
  "Pairing Rate", vdj_t_metrics$pairingRate,
  "Total Clone Types Detected", vdj_t_metrics$totalCloneTypes
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_t_tbl) <- c("", "")
vdj_t_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### Knee Plot

```{r eval = include_tcr}
cellNum <- as.numeric(vdj_t_metrics$cells)
vdj_t_kneeData <- read_tsv(params$VDJ_T_kneeOut, col_names = c("CB", "UMI"))
vdj_t_cells <- read_tsv(params$VDJ_T_cells)
vdj_t_knee <- vdj_t_kneeData %>% 
    mutate(
        cell = c(1:nrow(vdj_t_kneeData)),
        group = case_when(
            CB %in% vdj_t_cells$CB ~ "cell",
            TRUE ~ "background"
        )
    )
fig <- plot_ly(data = vdj_t_knee, x = ~cell, y = ~UMI, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'UMI: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log"),yaxis = list(type = "log"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()
```

Row {data-height=400}
----------

### Sequencing Stats

```{r eval = include_tcr}
##Number of Reads
##Reads With Valid Barcodes
##Sequencing Saturation
##Q30 Bases in CB+UMI
##Q30 Bases in RNA read
##Total VDJ Gene Fragments Detected
vdj_t_tbl <- tribble(
  ~term, ~value,
  "Number of Reads", vdj_t_metrics$totalRawReads,
  "Reads With Valid Barcodes", vdj_t_metrics$validBCreads,
  "Sequencing Saturation", vdj_t_metrics$saturation,
  "Q30 Bases in CB+UMI", vdj_t_metrics$q30InCBandUMI,
  "Q30 Bases in RNA read", vdj_t_metrics$q30InRNA
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_t_tbl) <- c("", "")
vdj_t_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### V(D)J Annotation

```{r eval = include_tcr}
##Cells with productive V-J spanning pair
##Cells with productive V-J spanning (IGK, IGH) pair
##Cells with productive V-J spanning (IGL, IGH) pair
##Cells with productive IGH contig
##Cells with productive IGK contig
##Cells with productive IGL contig
##Paired clonotype diversity
vdj_t_tbl <- tribble(
  ~term, ~value,
  "Median TRB/TRD UMIs per Cell", vdj_t_metrics$medianUMIsChain1,
  "Median TRA/TRG UMIs per Cell", vdj_t_metrics$medianUMIsChain2,
  "Cells with Productive TRB", vdj_t_metrics$cellsWithProductiveChainTRB,
  "Cells with Productive TRD", vdj_t_metrics$cellsWithProductiveChainTRD,
  "Cells with Productive TRA", vdj_t_metrics$cellsWithProductiveChainTRA,
  "Cells with Productive TRG", vdj_t_metrics$cellsWithProductiveChainTRG,
  "Cells with Productive TRATRB pair", vdj_t_metrics$cellsWithProductiveChainTRATRB,
  "Cells with Productive TRGTRD pair", vdj_t_metrics$cellsWithProductiveChainTRGTRD
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_t_tbl) <- c("", "")
vdj_t_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

Row
----------

### GEX Association

```{r eval = (include_tcr & include_gex)}
## Only plot cluster when there are more than 150 filtered cells
if(filtered_cells >= 150){
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
    ## extract data from seuratData
    tcr_cdrSeq <- read_tsv(params$VDJ_T_report) %>%
        select(`#barcode`, `chain1`, `chain2`)
    tcr_cdrSeq$chain1_cdr3 <- ""
    tcr_cdrSeq$chain2_cdr3 <- ""
    for(i in 1:nrow(tcr_cdrSeq)){
        tcr_cdrSeq$chain1_cdr3[i] <- unlist(strsplit(tcr_cdrSeq$chain1[i], ","))[6]
        tcr_cdrSeq$chain2_cdr3[i] <- unlist(strsplit(tcr_cdrSeq$chain2[i], ","))[6]
    }
    
    tcr_cells <- vdj_t_knee %>%
        filter(group == "cell") %>%
        pull(CB)
    tcr_data <- vdj_t_kneeData %>%
        filter(CB %in% tcr_cells) %>%
        left_join(tcr_cdrSeq, by = c("CB" = "#barcode"))
    clusterData2 <- clusterData %>% left_join(tcr_data, by = c("bc" = "CB")) %>%
        dplyr::select(bc, UMAP_1, UMAP_2, chain1_cdr3, chain2_cdr3) %>%
        dplyr::mutate(
            text = paste0(chain1_cdr3, "+", chain2_cdr3),
            type = case_when(
                !is.na(chain1_cdr3) & !is.na(chain2_cdr3) ~ "Detected",
                TRUE ~ "Not captured"
            )
        )

    plot_ly(clusterData2, x = ~UMAP_1, y= ~UMAP_2, color = ~type, 
            text = ~text,
            type = 'scatter', mode = 'markers',
            marker = list(opacity = 0.5),
            colors = c("#9370DB", "grey"),
            hovertemplate = 'CDR3: %{text}<extra><b>Type: %{fullData.name}</b></extra>') %>%
        config_plotly_fig() %>%
        toWebGL()
    
    ##clusterPlot %>% ggplotly() %>%
    ##    config(displaylogo = FALSE,
    ##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
    ##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

### Top 10 Clonetypes

```{r eval = include_tcr}
fig <- plot_ly(data = vdj_t_cloneType %>% filter(id<=10), x = ~id, y = ~CellCount, type = "bar")
fig <- layout(fig, showlegend = FALSE)
config_plotly_fig(fig)
```

Row {style="margin-right: 8px"}
----------

### Identified Clonetypes (top 1000) {style="width: 100% !important"}

```{r eval = include_tcr}
vdj_t_cloneType %>%
    filter(id <= 1000) %>%
    DT::datatable(rownames = FALSE, width = "100%",
                  options = list(scrollX = T, scrollY = F, pageLength = 10))
```

`r if(!include_tcr) {"\\end{comment}"} else {NULL}`

`r if(!include_bcr) {"\\begin{comment}"} else {NULL}`

VDJ-B {data-orientation=rows}
====================

<h2 style="margin-top: 0.5rem; font-weight: bold;">Sample: `r params$sampleName` (VDJ-B)</h2>


```{r include=FALSE, eval = include_bcr}
vdj_b_cloneType <- read_tsv(params$VDJ_B_cloneType) %>%
    arrange(desc(CellCount))
vdj_b_cloneType <- vdj_b_cloneType %>%
    mutate(id = 1:nrow(vdj_b_cloneType),
           Frequency = scales::label_percent(0.01)(Frequency))
```

Row
----------

### Estimated Number of Cells


```{r include = FALSE, eval = include_bcr}
vdj_b_metrics <- fromJSON(params$VDJ_B_metrics)
fromJSON(params$VDJ_B_metrics) %>% 
    as.data.frame %>% t() %>% as.data.frame() %>% 
    rownames_to_column("term") %>% 
    write_tsv(paste0(params$sampleName, "_VDJ-B.metrics.tsv"), col_names=FALSE)
```

```{r eval = include_bcr}
valueBox(vdj_b_metrics$cells %>% as.numeric() %>% label_comma()(), icon = "fa-circle-notch", color = "info")
```

### Mean Reads per Cell

```{r eval = include_bcr}
valueBox(vdj_b_metrics$meanReadsPerCell %>% as.numeric() %>% round %>% label_comma()(), icon = "fa-chart-bar", color = "primary")
```

### Total Clonetypes Discovered

```{r eval = include_bcr}
valueBox(nrow(vdj_b_cloneType) %>% label_comma()(), icon = "fa-barcode", color = "warning")
```

Row
----------

### VDJ Stats

```{r eval = include_bcr}
##Estimated Number of Cells
##Reads in Cells Mapped to Gene
##Fraction of Reads in Cells
##Mean Reads per Cell
##Median Reads per Cell
##UMIs in Cells
##Mean UMI per Cell
##Median UMI per Cell
##Mean Gene per Cell
##Median Gene per Cell

fractionAssembled <- as.numeric(vdj_b_metrics$totalReadsInCell)/(as.numeric(vdj_b_metrics$totalRawReads)*as.numeric(vdj_b_metrics$validBCreads))
fractionAssembled <- as.character(fractionAssembled)

vdj_b_tbl <- tribble(
  ~term, ~value,
  "Estimated Number of Cells", vdj_b_metrics$cells,
  "Reads in Cells Mapped to VDJ Gene", vdj_b_metrics$totalReadsInCell,
  "Fraction of Reads Used in Assembly", fractionAssembled,
  "Mean Reads per Cell", vdj_b_metrics$meanReadsPerCell,
  "Median Reads per Cell", vdj_b_metrics$medianReadsPerCell,
  "Mean UMIs per Cell", vdj_b_metrics$meanUMIsPerCell,
  "Median UMIs per Cell", vdj_b_metrics$medianUMIsPerCell,
  "Pairing Rate", vdj_b_metrics$pairingRate,
  "Total Clone Types Detected", vdj_b_metrics$totalCloneTypes
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_b_tbl) <- c("", "")
vdj_b_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### Knee Plot

```{r eval = include_bcr}
cellNum <- as.numeric(vdj_b_metrics$cells)
vdj_b_kneeData <- read_tsv(params$VDJ_B_kneeOut, col_names = c("CB", "UMI"))
vdj_b_cells <- read_tsv(params$VDJ_B_cells)
vdj_b_knee <- vdj_b_kneeData %>% 
    mutate(
        cell = c(1:nrow(vdj_b_kneeData)),
        group = case_when(
            CB %in% vdj_b_cells$CB ~ "cell",
            TRUE ~ "background"
        )
    )
fig <- plot_ly(data = vdj_b_knee, x = ~cell, y = ~UMI, type = 'scatter', mode = 'lines',
               color = ~group, colors = c('grey', '#002366'),
               hovertemplate = 'UMI: %{y}<extra>%{fullData.name}</extra>',
               line = list(width = 3),
               height = 400, width = 360)
fig <- layout(fig, xaxis = list(type = "log"),yaxis = list(type = "log"), showlegend = FALSE)
config_plotly_fig(fig) %>% toWebGL()
```

Row {data-height=400}
----------

### Sequencing Stats

```{r eval = include_bcr}
##Number of Reads
##Reads With Valid Barcodes
##Sequencing Saturation
##Q30 Bases in CB+UMI
##Q30 Bases in RNA read
##Total VDJ Gene Fragments Detected

vdj_b_tbl <- tribble(
  ~term, ~value,
  "Number of Reads", vdj_b_metrics$totalRawReads,
  "Reads With Valid Barcodes", vdj_b_metrics$validBCreads,
  "Sequencing Saturation", vdj_b_metrics$saturation,
  "Q30 Bases in CB+UMI", vdj_b_metrics$q30InCBandUMI,
  "Q30 Bases in RNA read", vdj_b_metrics$q30InRNA
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_b_tbl) <- c("", "")
vdj_b_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

### V(D)J Annotation

```{r eval = include_bcr}
##Cells with productive V-J spanning pair
##Cells with productive V-J spanning (IGK, IGH) pair
##Cells with productive V-J spanning (IGL, IGH) pair
##Cells with productive IGH contig
##Cells with productive IGK contig
##Cells with productive IGL contig
##Paired clonotype diversity
vdj_b_tbl <- tribble(
  ~term, ~value,
  "Median IGH UMIs per Cell", vdj_b_metrics$medianUMIsChain1,
  "Median IGK/L UMIs per Cell", vdj_b_metrics$medianUMIsChain2,
  "Cells with Productive IGH", vdj_b_metrics$cellsWithProductiveChainIGH,
  "Cells with Productive IGK", vdj_b_metrics$cellsWithProductiveChainIGK,
  "Cells with Productive IGL", vdj_b_metrics$cellsWithProductiveChainIGL,
  "Cells with Productive IGKIGH pair", vdj_b_metrics$cellsWithProductiveChainIGKIGH,
  "Cells with Productive IGLIGH pair", vdj_b_metrics$cellsWithProductiveChainIGLIGH
) %>%
  mutate(value = as.numeric(value)) %>%
  mutate(value=case_when(value <= 1 ~ label_percent(0.01)(value),
                         value > 1 ~ label_comma()(round(value))))
names(vdj_b_tbl) <- c("", "")
vdj_b_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

Row
----------

### GEX Association

```{r eval = (include_bcr & include_gex)}
## Only plot cluster when there are more than 150 filtered cells
if(filtered_cells >= 150){
    ##clusterPlot <- DimPlot(object=seuratData, reduction="umap", label=TRUE, repel=TRUE)
    bcr_cdrSeq <- read_tsv(params$VDJ_B_report) %>%
        select(`#barcode`, `chain1`, `chain2`)
    bcr_cdrSeq$chain1_cdr3 <- ""
    bcr_cdrSeq$chain2_cdr3 <- ""
    for(i in 1:nrow(bcr_cdrSeq)){
        bcr_cdrSeq$chain1_cdr3[i] <- unlist(strsplit(bcr_cdrSeq$chain1[i], ","))[6]
        bcr_cdrSeq$chain2_cdr3[i] <- unlist(strsplit(bcr_cdrSeq$chain2[i], ","))[6]
    }
    ## extract data from seuratData
    bcr_cells <- vdj_b_knee %>%
        filter(group == "cell") %>%
        pull(CB)
    bcr_data <- vdj_b_kneeData %>%
        filter(CB %in% bcr_cells) %>%
        left_join(bcr_cdrSeq, by = c("CB" = "#barcode"))
    clusterData2 <- clusterData %>% left_join(bcr_data, by = c("bc" = "CB")) %>%
        dplyr::select(bc, UMAP_1, UMAP_2, chain1_cdr3, chain2_cdr3) %>%
        dplyr::mutate(
            text = paste0(chain1_cdr3, "+", chain2_cdr3),
            type = case_when(
                !is.na(chain1_cdr3) & !is.na(chain2_cdr3) ~ "Detected",
                TRUE ~ "Not captured"
            )
        )

    plot_ly(clusterData2, x = ~UMAP_1, y= ~UMAP_2, color = ~type, 
            text = ~text,
            type = 'scatter', mode = 'markers',
            marker = list(opacity = 0.5),
            colors = c("#9370DB", "grey"),
            hovertemplate = 'CDR3: %{text}<extra><b>Type: %{fullData.name}</b></extra>') %>%
        config_plotly_fig() %>%
        toWebGL()
    
    ##clusterPlot %>% ggplotly() %>%
    ##    config(displaylogo = FALSE,
    ##           modeBarButtonsToRemove = c('zoom', 'pan', 'select', 'zoomIn', 'zoomOut', 'autoScale',
    ##                                      'hoverClosestCartesian', 'hoverCompareCartesian')) %>% toWebGL()
}else{
    cat("Too few cells: ", filtered_cells, "(filtered by nFeature >= 200)\n")
}
```

### Top 10 Clonetypes

```{r eval = include_bcr}
fig <- plot_ly(data = vdj_b_cloneType %>% filter(id<=10), x = ~id, y = ~CellCount, type = "bar")
fig <- layout(fig, showlegend = FALSE)
config_plotly_fig(fig)
```

Row {style="margin-right: 8px"}
----------

### Identified Clonetypes (top 1000) {style="width: 100% !important"}

```{r eval = include_bcr}
vdj_b_cloneType %>%
    filter(id <= 1000) %>%
    DT::datatable(rownames = FALSE, width = "100%",
                  options = list(scrollX = T, scrollY = F, pageLength = 10))
```

`r if(!include_bcr) {"\\end{comment}"} else {NULL}`

Running Info {data-orientation=rows}
====================

Row
----------

### Running Information

```{r}
version_info <- fromJSON(params$version_json)
version_tbl <- tribble(
  ~term, ~value,
  "Sample ID", params$sampleName,
  "Pipeline Version", version_info$pipeline_version,
  "Reference Dir", version_info$referenceDir,
  "Reference GTF", version_info$referenceGTF,
  "STAR Version", version_info$STAR_version,
  "Whitelist Matching", version_info$soloCBmatchWLtype,
  "UMIfiltering", version_info$soloUMIfiltering,
  "UMIdedup", version_info$soloUMIdedup,
  "Cell Calling", version_info$soloCellFilter,
  "includeIntron", version_info$includeIntron,
  "includeMultiReads", version_info$includeMultiReads
)
names(version_tbl) <- c("", "")
version_tbl %>% 
    kbl() %>% 
    kable_paper("hover", html_font = "arial, helvetica, sans-serif")
```

```{js}
// set knee plot margin to auto
$( document ).ready(function() {
    document.querySelectorAll("div.plot-container > div.svg-container").forEach(el => el.style.margin = "auto");
});
```